File: 112-Path-Sum.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if root is None:
            return False

        stuck = [[root, root.val]]

        while len(stuck) >= 1:
            curElem = stuck.pop(0)
            curNode = curElem[0]
            curSum = curElem[1]

            # Leaf
            if curNode.left is None and curNode.right is None:
                if curSum == sum:
                    return True
            # Not Leaf
            else:
                if curNode.left is not None:
                    stuck.insert(0, [curNode.left, curSum + curNode.left.val])
                if curNode.right is not None:
                    stuck.insert(
                        0, [curNode.right, curSum + curNode.right.val])

        return False

# Sample Testcase:
#   Input:
#       [5,4,8,11,null,13,4,7,2,null,null,null,1]
#       22
#   Output:
#       True

# Performance Result:
#   Coding Time: 00:11:24
#   Time Complexity: O(n)
#   Space Complexity: O(n)
#   Runtime: 52 ms, faster than 63.25% of Python3
#       online submissions for Path Sum.
#   Memory Usage: 15.7 MB, less than 5.45% of Python3
#       online submissions for Path Sum.

File: 112-Path-Sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type NextNode struct {
    Node *TreeNode
    Total int
}

func hasPathSum(root *TreeNode, targetSum int) bool {
    if root == nil {
        return false
    }

    nextNodes := []NextNode{{Node: root, Total: root.Val}}
    for len(nextNodes) > 0 {
        targetNode := nextNodes[0]
        if targetNode.Node.Left == nil && targetNode.Node.Right == nil && targetNode.Total == targetSum {
            return true
        }

        nextNodes = nextNodes[1:]
        if targetNode.Node.Left != nil {
            total := targetNode.Total + targetNode.Node.Left.Val
            nextNodes = append(nextNodes, NextNode{Node: targetNode.Node.Left, Total: total})
        }
        if targetNode.Node.Right != nil {
            total := targetNode.Total + targetNode.Node.Right.Val
            nextNodes = append(nextNodes, NextNode{Node: targetNode.Node.Right, Total: total})
        }
    }

    return false
}

// Performance Result:
//   Coding Time: 00:27:31
//   Time Complexity: O(log n)
//   Space Complexity: O(log n)
//   Runtime: 0 ms, faster than 100.00%
//   Memory Usage: 5.42 MB, less than 11.98%
